### PR TITLE
Fix types, introduce type tests

### LINKED ISSUE #2558: Typing changes in 8.1.4 cause failure of mypy check
Click's simple example https://github.com/pallets/click#a-simple-example doesn't pass mypy 1.4.1 since 8.1.4.  With click 8.1.3 the example did pass. This affects not only the simple example but also existing projects that use click. I ran mypy on one of my projects and everything passed, ran it 30 minutes later and saw failures out of the blue. Click 8.1.4 was published to PyPI between my checks.

To reproduce:

* Save the simple example as click.py
* `pip install mypy==1.4.1 click==8.1.3`
* `mypy click.py`

This is the output:

```
click.py:3: error: Argument 1 has incompatible type "Callable[[Any, Any], Any]"; expected <nothing>  [arg-type]
click.py:12: error: <nothing> not callable  [misc]
```

Environment:

- Python version: 3.11
- Click version: 8.1.4

### PR DESCRIPTION
**Pull request by Tinche**
Fix signatures, introduce simple tests.

I followed the Flask model for typing tests. Initially I was going to replicate the attrs model (https://github.com/python-attrs/attrs/blob/main/tests/test_mypy.yml) but maybe this is good enough to start?

Also introduced Pyright tests, again copied from attrs.

### CODE REVIEW

File: src/click/decorators.py [hunk 1]
```diff
@@ -150,16 +150,12 @@ def command(
...
-# However, mypy thinks this doesn't fit the overloaded function. Pyright does
-# accept that spelling, and the following work-around makes pyright issue a
-# warning that CmdType could be left unsolved, but mypy sees it as fine. *shrug*
+# variant: name omitted, cls _must_ be a keyword argument, @command(cls=CommandCls, ...)
 @t.overload
 def command(
     name: None = None,
-    cls: t.Type[CmdType] = ...,
+    *,
+    cls: t.Type[CmdType],
```
Comment (layday) [inline src/click/decorators.py:158]: I think the same change should be made to `group`'s third overload.

**Comment by davidism** [comment]
We don't have CI for this yet, but check `pyright --verifytypes click` to make sure pyright can at least understand the types.

Should this be used instead of #2559?

**Comment by Tinche** [comment]
> Should this be used instead of https://github.com/pallets/click/pull/2559?

I've added `help_option` to the typing tests, so this is now a strict superset. So pretty much yeah.

**Comment by Tinche** [comment]
> We were also using pyright --verifytypes click, not sure what the exact difference is but might want to run both.

I've added it. Looks like it prints out a bunch of stuff and succeeds.

```
pyright: commands[0]> pyright --verifytypes click
Module name: "click"
Package directory: "/home/tin/pg/click/.tox/pyright/lib/python3.11/site-packages/click"
Module directory: "/home/tin/pg/click/.tox/pyright/lib/python3.11/site-packages/click"
Path of py.typed file: "/home/tin/pg/click/.tox/pyright/lib/python3.11/site-packages/click/py.typed"

Public modules: 12
   click
   click.core
   click.decorators
   click.exceptions
   click.formatting
   click.globals
   click.parser
   click.shell_completion
   click.termui
   click.testing
   click.types
   click.utils

Symbols used in public interface:

Symbols exported by "click": 646
  With known type: 646
  With ambiguous type: 0
  With unknown type: 0
  Functions without docstring: 98
  Functions without default param: 22
  Classes without docstring: 13

Other symbols referenced but not exported by "click": 247
  With known type: 247
  With ambiguous type: 0
  With unknown type: 0

Type completeness score: 100%

Completed in 1.389sec
```

File: src/click/decorators.py (hunk 1 above)
Comment (Tinche) [reply src/click/decorators.py:157]: Sure, but I think it's going to be up to the community to fill out the tests and fix the fallout. This PR is mainly to set up the infrastructure for it.

**Comment by ofek** [comment]
Is this ready to be merged?

**Comment by davidism** [comment]
Yes, I just need to find some time, busy with regular job.

**Comment by davidism** [comment]
I've switched from installing and using nodenv directly to installing the pyright package from PyPI. This allows us to pin the pyright version the same as mypy. I prefer to pin these tools because new versions regularly cause our tests to fail, which is confusing to contributors making PRs. (That said, the failures are usually nice things like being able to remove ignore comments.) Everything seems to work still, but let me know if there was a specific reason for using nodenv directly.

**Comment by Tinche** [comment]
Cool, thanks for the tip. I've actually borrowed this code from attrs, I'm not a pyright expert or anything. Maybe @hynek has an opinion?

**Pull request merged**